src/index.js

```diff
@@ -279,22 +282,28 @@ class Bitswap {
   async putMany (blocks) { // eslint-disable-line require-await
     const self = this
 
-    return this.blockstore.putMany(async function * () {
+    // Add any new blocks to the blockstore
+    const newBlocks = []
+    await this.blockstore.putMany(async function * () {
       for await (const block of blocks) {
         if (await self.blockstore.has(block.cid)) {
           continue
         }
 
         yield block
-
-        self.notifications.hasBlock(block)
-        self.engine.receivedBlocks([block.cid])
-        // Note: Don't wait for provide to finish before returning
-        self.network.provide(block.cid).catch((err) => {
-          self._log.error('Failed to provide: %s', err.message)
-        })
+        newBlocks.push(block)
       }
     }())
+
+    // Notify listeners that we have received the new blocks
+    for (const block of newBlocks) {
+      self.notifications.hasBlock(block)
+      self.engine.receivedBlocks([block])
```

Comment:
No need to `self` anymore, but why did this get pulled out here? If it stays, could you pass `newBlocks` to       `self.engine.receivedBlocks(newBlocks)` and not have to call it multiple times?

Reply:
It's a little deceptive: the async iterator passed to `putMany()` yields blocks that should be put (it doesn't yield blocks that *have already* been put). So we actually need to perform post-put actions after the whole `putMany()` call has completed.

You're right about the `receivedBlocks()` call, it's more efficient to pass all blocks at once, I've fixed 👍